bin/jetstream.php

```diff
@@ -0,0 +1,855 @@
+				$profiles[$follow->did] = $follow;
+				$follows[$follow->did]  = Contact::SHARING;
+			}
+			$cursor = $data->cursor ?? '';
+		} while (!empty($data->follows) && !empty($data->cursor));
+
+		$cursor = '';
+
+		do {
+			$parameters = [
+				'actor'  => $this->getUserDid($uid),
+				'limit'  => 100,
+				'cursor' => $cursor
+			];
+
+			$data = $this->XRPCGet('app.bsky.graph.getFollowers', $parameters);
+
+			foreach ($data->followers ?? [] as $follow) {
+				$profiles[$follow->did] = $follow;
+				$follows[$follow->did]  = ($follows[$follow->did] ?? 0) | Contact::FOLLOWER;
+			}
+			$cursor = $data->cursor ?? '';
+		} while (!empty($data->followers) && !empty($data->cursor));
+
+		foreach ($contacts as $contact) {
+			if (empty($follows[$contact['url']])) {
+				Contact::update(['rel' => Contact::NOTHING], ['id' => $contact['id']]);
+			}
+		}
+
+		foreach ($follows as $did => $rel) {
+			$contact = bluesky_get_contact($profiles[$did], $uid, $uid);
+			if ($contact['rel'] != $rel) {
+				Contact::update(['rel' => $rel], ['id' => $contact['id']]);
+			}
+		}
+		Logger::info('Sync contacts for user - done', ['uid' => $uid]);
+	}
+
+	private function setOptions(\WebSocket\Client $client)
+	{
+		$uids = $this->getUids();
+		if (empty($uids)) {
+			return;
+		}
+
+		foreach ($uids as $uid) {
+			$this->syncContacts($uid);
+		}
+
+		$contacts = Contact::selectToArray(['uid', 'url'], ['uid' => $uids, 'network' => Protocol::BLUESKY, 'rel' => [Contact::FRIEND, Contact::SHARING]]);
+
+		$self = [];
+		foreach ($uids as $uid) {
+			$did        = $this->getUserDid($uid);
+			$contacts[] = ['uid' => $uid, 'url' => $did];
+			$self[$did] = $uid;
+		}
+		$this->self = $self;
+
+		$uids = [];
+		foreach ($contacts as $contact) {
+			$uids[$contact['url']][] = $contact['uid'];
+		}
+		$this->uids = $uids;
+
+		if (count($uids) > 10000) {
+			$uids = [];
+			Logger::warning('More than 10,000 contacts - switch to firehose');
+		}
+
+		$update = [
+			'type'    => 'options_update',
+			'payload' => [
+				'wantedCollections'   => [],
+				'wantedDids'          => array_keys($uids),
+				'maxMessageSizeBytes' => 1000000
+			]
+		];
+		$client->send(json_encode($update));
+	}
+
+	private function getUids(): array
+	{
+		$uids         = [];
+		$abandon_days = intval(DI::config()->get('system', 'account_abandon_days'));
+		if ($abandon_days < 1) {
+			$abandon_days = 0;
+		}
+
+		$abandon_limit = date(DateTimeFormat::MYSQL, time() - $abandon_days * 86400);
+
+		$pconfigs = DBA::selectToArray('pconfig', [], ["`cat` = ? AND `k` = ? AND `v`", 'bluesky', 'import']);
+		foreach ($pconfigs as $pconfig) {
+			if (empty($this->getUserDid($pconfig['uid']))) {
+				continue;
+			}
+
+			if ($abandon_days != 0) {
+				if (!DBA::exists('user', ["`uid` = ? AND `login_date` >= ?", $pconfig['uid'], $abandon_limit])) {
+					continue;
+				}
+			}
+			$uids[] = $pconfig['uid'];
+		}
+		return $uids;
+	}
+
+	private function route(stdClass $data)
```

Comment:
Yes, please move this class into a separate file. Next we can create helper classes and value objects to prevent [moving around primitives](https://refactoring.guru/smells/primitive-obsession).

Reply:
So we would have some kind of `JetstreamDataTransferObject` class, where the `stdClass $data` is the input, some basic checks/transformations and then we can use the `$jetstreamDTO` as input for the `JetstreamDataProcessor` - which basically creates/deletes/updates content (Item, Posts, ...)

Reply:
@Art4 I will at first move this into a single class then we can see how to continue. I guess that there are fancy ways in splitting this. But keep in mind: Although I'm a professional coder now for more than 30 years, I never worked professionally with PHP and I don't work with classes at work. At work we use a language that doesn't know classes. There are several places in the current code that are too abstract for me to even find the relevant code, not to mention to extend it.

Reply:
@annando You have my full respect for your work! I would be glad if I can assist you to create a more robust code base. I'm doing refactoring PHP code for over 15 years and I have experience in introducing improvements and changes gradually so that all employees can adapt slowly and see the benefit of the changes.

Reply:
Be prepared that I really need to be convinced and that I only accept things where I can see the benefit :grin: I don't accept a "we're doing it because everyone else is doing it" - only a "We do it because the benefit is ...".

At work (I'm a lead developer) I often get into intense discussions with one of my colleagues (his job description is "engineer") about specific designs of the applications we build. We are both stubborn - but in the end we find a solution that we both like.